workflows/src/main/kotlin/com/r3/corda/lib/ci/RequestKeyFlow.kt

```diff
@@ -0,0 +1,93 @@
+package com.r3.corda.lib.ci
+
+import co.paralleluniverse.fibers.Suspendable
+import net.corda.core.contracts.UniqueIdentifier
+import net.corda.core.crypto.SecureHash
+import net.corda.core.crypto.SignedData
+import net.corda.core.flows.FlowException
+import net.corda.core.flows.FlowLogic
+import net.corda.core.flows.FlowSession
+import net.corda.core.identity.AnonymousParty
+import net.corda.core.identity.Party
+import net.corda.core.serialization.deserialize
+import net.corda.core.utilities.ProgressTracker
+import net.corda.core.utilities.toBase58String
+import net.corda.core.utilities.unwrap
+import java.security.PublicKey
+import java.util.*
+
+/**
+ * This flow registers a mapping in the [net.corda.core.node.services.IdentityService] between a [PublicKey] and a [Party]. It can generate a new key
+ * pair for a given [UUID] and register the new key mapping, or a known [PublicKey] can be supplied to the flow which will register
+ * a mapping between this key and the requesting party.
+ *
+ * The generation of the signed [SignedOwnershipClaim] is delegated to the counter-party that sends the signed mapping back to
+ * the requesting node. The requesting node verifies the signature of the signed mapping matches that of the counter-party
+ * before registering the mapping in the [net.corda.core.node.services.IdentityService].
+ */
+class RequestKeyFlow
+private constructor(
+        private val session: FlowSession,
+        private val uuid: UUID,
+        private val key: PublicKey?) : FlowLogic<SerializedSignedOwnershipClaim<SignedOwnershipClaim>>() {
+    constructor(session: FlowSession, uuid: UUID) : this(session, uuid, null)
+    constructor(session: FlowSession, key: PublicKey) : this(session, UniqueIdentifier().id, key)
+
+    companion object {
+        object REQUESTING_KEY : ProgressTracker.Step("Requesting a public key")
+        object VERIFYING_KEY : ProgressTracker.Step("Verifying counterparty's signature")
+        object KEY_VERIFIED : ProgressTracker.Step("Signature is correct")
+
+        @JvmStatic
+        fun tracker(): ProgressTracker = ProgressTracker(REQUESTING_KEY, VERIFYING_KEY, KEY_VERIFIED)
+    }
+
+    override val progressTracker = tracker()
+
+    @Suspendable
+    @Throws(FlowException::class)
+    override fun call(): SerializedSignedOwnershipClaim<SignedOwnershipClaim> {
+        progressTracker.currentStep = REQUESTING_KEY
+        val nonce = OwnershipClaim(SecureHash.randomSHA256().bytes)
```

Comment:
I am not sure what the purpose of this nonce is supposed to be? A secure random number gives good guarantee that this really is a nonce (i.e. a unique / non-repeating number) - but nothing is checking it afaict ?